src/components/data/DataTable.js

```diff
@@ -0,0 +1,190 @@
+import { useContext, useEffect } from 'preact/hooks';
+
+import { useSyncedRef } from '../../hooks/use-synced-ref';
+import { downcastRef } from '../../util/typing';
+import { useArrowKeyNavigation } from '../../hooks/use-arrow-key-navigation';
+
+import { SpinnerSpokesIcon } from '../icons';
+import ScrollContext from '../data/ScrollContext';
+import Table from './Table';
+import TableHead from './TableHead';
+import TableBody from './TableBody';
+import TableRow from './TableRow';
+import TableCell from './TableCell';
+
+/**
+ * @typedef {import('../../types').CompositeProps} CommonProps
+ * @typedef {Omit<import('preact').JSX.HTMLAttributes<HTMLElement>, 'size'|'rows'|'role'|'loading'>} HTMLAttributes
+ * @typedef {import('./TableContext').TableInfo} TableInfo
+ * @typedef {import('preact').ComponentChildren} Children
+ *
+ * @typedef Column
+ * @prop {string} field
+ * @prop {string} label
+ * @prop {string} [classes]
+ *
+ * @typedef {Record<string, any>} Row
+ *
+ * @typedef {( row:Row, field:string) => Children|null} RenderItemCallback
+ * @typedef {( row: Row ) => void } RowCallback
+ *
+ * @typedef DataTableProps
+ * @prop {Column[]} columns
+ * @prop {Row[]} rows
+ * @prop {string} title
+ * @prop {Row|null} [selectedRow]
+ * @prop {boolean} [loading=false]
+ * @prop {RowCallback} [onConfirmRow] - Callback when a row is "confirmed" by
+ *   double-click or pressing "Enter"
+ * @prop {RowCallback} [onSelectRow] - Callback when a row is "selected" by
+ *   focus or click
+ * @prop {RenderItemCallback} [renderItem] - Callback to render an individual table cell
+ * @prop {Children} [emptyMessage] - Content to render if rows is empty (and not
+ *   in a loading state)
+ */
+
+/**
+ * An interactive table of rows and columns with a sticky header.
+ *
+ * @param {CommonProps & DataTableProps & HTMLAttributes} props
+ */
+const DataTableNext = function DataTable({
+  children,
+  elementRef,
+
+  columns = [],
+  rows = [],
+  title,
+  selectedRow,
+  loading = false,
+  renderItem,
+  onSelectRow,
+  onConfirmRow,
+  emptyMessage,
+
+  ...htmlAttributes
+}) {
+  const tableRef = useSyncedRef(elementRef);
+  const scrollContext = useContext(ScrollContext);
+
+  useArrowKeyNavigation(tableRef, {
+    selector: 'tbody tr',
+    horizontal: true,
+    vertical: true,
+  });
+
+  const noContent = loading || (!rows.length && emptyMessage);
+  const fields = columns.map(column => column.field);
+
+  /**
+   * @param {Row} row
+   * @param {string} field
+   */
+  function formatDataRowItem(row, field) {
+    return renderItem?.(row, field) ?? row[field];
+  }
+
+  /** @param {Row} row */
+  function selectRow(row) {
+    onSelectRow?.(row);
+  }
+
+  /** @param {Row} row */
+  function confirmRow(row) {
+    onConfirmRow?.(row);
+  }
+
+  /** @param {KeyboardEvent} event */
+  function handleKeyDown(event, row) {
```

Comment:
Param type for `row` is missing here. The `noImplicitAny` setting is currently off for this project in tsconfig.json, so TS doesn't complain. We should address that at some point.